### PR TITLE
sync_sso() adds its own "custom." prefix. This results in "custom.custom.field" in the payload

### PR DESCRIPTION
**Pull request by vkozyrev**
* `DiscourseApi::SingleSignOn.custom_fields` adds its own `custom.`	prefix before sending the payload
* `sync_sso` expected the	params to have the prefix in the params	
* This results in `custom.custom.field` in the payload

Also added some overkill tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,7 +52,8 @@ client.sync_sso(                                #=> Synchronizes the SSO record
   name: "Test Name",
   username: "test_name",
   email: "name@example.com",
-  external_id: "2"
+  external_id: "2",
+  "custom.field_1" => "Custom field"
```
Comment (vkozyrev) [inline README.md]: I think this is better than `:"custom.field_1"`

**Comment by oblakeerickson** [comment]
@vkozyrev thank you for making this pr. Just letting you know I've seen it and will review it when I get a chance.

**Comment by oblakeerickson** [comment]
After diving into the code here for a bit I think we should just remove the prepending of `.custom` inside of the unsigned_payload method:

https://github.com/discourse/discourse_api/blob/a2ff67b0f267ca6d1594c38c6f4baca1e46978d4/lib/discourse_api/single_sign_on.rb#L104

I don't know why it is there. Especially since you have to have `custom` prepended already in the sync_sso method

https://github.com/discourse/discourse_api/blob/a2ff67b0f267ca6d1594c38c6f4baca1e46978d4/lib/discourse_api/api/sso.rb#L18

This means you will have to update your PR quite a bit, but I think that is the proper place to make this fix.

File: README.md (hunk 1 above)
Comment (oblakeerickson) [inline README.md]: :+1:

File: lib/discourse_api/api/sso.rb [hunk 2]
```diff
@@ -15,8 +15,8 @@ def sync_sso(params = {})
         sso.avatar_force_update = params[:avatar_force_update] === true
         sso.add_groups = params[:add_groups]
         sso.remove_groups = params[:remove_groups]
-        params.keys.select { |key| key.to_s.start_with?("custom") }.each do |custom_key|
-          sso.custom_fields[custom_key] = params[custom_key]
+        params.keys.select { |key| key.to_s.start_with?('custom.') }.each do |custom_key|
```
Comment (oblakeerickson) [inline lib/discourse_api/api/sso.rb]: :+1:

File: lib/discourse_api/api/sso.rb [hunk 3]
```diff
@@ -15,8 +15,8 @@ def sync_sso(params = {})
...
+          sso.custom_fields[custom_key.sub('custom.', '')] = params[custom_key]
```
Comment (oblakeerickson) [inline lib/discourse_api/api/sso.rb]: See my comment on your PR. We should fix this inside of the SingleSignOn class

File: spec/discourse_api/api/sso_spec.rb [hunk 4]
```diff
@@ -3,13 +3,60 @@
...
+      username: 'some_user',
+      email: 'some@email.com',
+      external_id: 'abc',
+      suppress_welcome_message: false,
+      avatar_url: 'https://www.website.com',
+      title: 'ruby',
+      avatar_force_update: false,
+      add_groups: ['a', 'b'],
+      remove_groups: ['c', 'd'],
+      'custom.custom.field_1' => 'potato',
```
Comment (oblakeerickson) [inline spec/discourse_api/api/sso_spec.rb]: This shouldn't be necessary anymore

**Comment by vkozyrev** [comment]
I'll update the PR accordingly sometime this week. 

This will be a breaking change for anyone using the `SingleSignOn` class directly.
Something like `sso.custom_fields['role'] = my_thing.role` will have to be updated to `sso.custom_fields['custom.role']`.

I don't think theres other places in the code that will be affected by this, do we need any sort of deprecation messaging @oblakeerickson ?

**Comment by vkozyrev** [comment]
I've updated the PR @oblakeerickson . Again be aware anyone using `custom_fields` with `DiscourseApi::SingleSignOn` directly will have issues.

File: lib/discourse_api/single_sign_on.rb [hunk 5]
```diff
@@ -99,11 +99,7 @@ def unsigned_payload
        payload[k] = val
       end
 
-      if @custom_fields
-        @custom_fields.each do |k, v|
-          payload["custom.#{k}"] = v.to_s
-        end
-      end
+      @custom_fields.each { |k, v| payload[k] = v.to_s } if @custom_fields
```
Comment (oblakeerickson) [inline lib/discourse_api/single_sign_on.rb]: Sorry for changing my mind on this, but I think we should just maintain backwards compatibility here. Let's check if it starts with `custom.` and if not add it. I still don't understand why people would use the single_sign_on class directly over api/sso but let's just not break things here.

Comment (vkozyrev) [reply lib/discourse_api/single_sign_on.rb]: Will do in the next few days 👍

Comment (oblakeerickson) [reply lib/discourse_api/single_sign_on.rb:102]: Is this still something you need/want to work on?

Comment (vkozyrev) [reply lib/discourse_api/single_sign_on.rb:102]: Hey, yea I'll do it this week. Sorry for the delay, moved to a non-discourse project at work.

Comment (vkozyrev) [reply lib/discourse_api/single_sign_on.rb:102]: Next week 😬 

I'm in the process of updating the gems on my project to upgrade to Rails 6. Going to open a separate PR to update to Faraday 1.0 as well. Got a few other gem issues I have to get through first before circling back to this.

File: lib/discourse_api/api/sso.rb [hunk 6]
```diff
@@ -3,24 +3,13 @@ module DiscourseApi
...
-        sso.remove_groups = params[:remove_groups]
-        params.keys.select { |key| key.to_s.start_with?("custom") }.each do |custom_key|
-          sso.custom_fields[custom_key] = params[custom_key]
-        end
+        sso_secret = params.delete(:sso_secret)
+        _sso_url = params.delete(:sso_url)
+        querystring = URI.encode_www_form(params)
+
+        sso = DiscourseApi::SingleSignOn.unsigned_parse(querystring, sso_secret)
+
```
Comment (vkozyrev) [inline lib/discourse_api/api/sso.rb:11]: I will update unsigned_parse to take hash

Comment (vkozyrev) [reply lib/discourse_api/api/sso.rb:11]: This way we don't have to add SSO params to the sync_sso method every time we add them to DiscourseApi::SingleSignOn::ACCESSORS

**Comment by vkozyrev** [comment]
let(:params) do
    {
      sso_secret: 'abc',
      sso_url: 'www.google.com',
      name: 'Some User',
      username: 'some_user',
      email: 'some@email.com',
      external_id: 'abc',
      suppress_welcome_message: false,
      avatar_url: 'https://www.website.com',
      title: 'ruby',
      avatar_force_update: false,
      add_groups: ['a', 'b'],
      remove_groups: ['c', 'd'],
      'custom.field_1' => 'tomato',
      'field_2' => 'potato'
    }
  end 
    
Master sso.unsigned_payload output:
`name=Some+User&username=some_user&email=some%40email.com&avatar_url=https%3A%2F%2Fwww.website.com&avatar_force_update=false&external_id=abc&suppress_welcome_message=false&title=ruby&add_groups=a&add_groups=b&remove_groups=c&remove_groups=d&custom.custom.field_1=tomato`

My sso.unsigned_payload output:
`name=Some+User&username=some_user&email=some%40email.com&avatar_url=https%3A%2F%2Fwww.website.com&avatar_force_update=false&external_id=abc&suppress_welcome_message=false&title=ruby&add_groups=a&add_groups=b&remove_groups=c&remove_groups=d&custom.custom.field_1=tomato&custom.field_2=potato`

File: lib/discourse_api/single_sign_on.rb [hunk 7]
```diff
@@ -25,6 +26,36 @@ def self.sso_url
       raise RuntimeError, "sso_url not implemented on class, be sure to set it on instance"
     end
```
Comment (vkozyrev) [inline lib/discourse_api/single_sign_on.rb:28]: Initialization is similar to `parse`, except we don't need to stringily the keys

File: spec/discourse_api/api/sso_spec.rb [hunk 8]
```diff
@@ -4,9 +4,51 @@
...
+      name: 'Some User',
+      username: 'some_user',
+      email: 'some@email.com',
+      external_id: 'abc',
+      suppress_welcome_message: false,
+      avatar_url: 'https://www.website.com',
+      title: 'ruby',
+      avatar_force_update: false,
+      add_groups: ['a', 'b'],
+      remove_groups: ['c', 'd'],
```
Comment (vkozyrev) [inline spec/discourse_api/api/sso_spec.rb:20]: old + new `custom_fields` formats

File: spec/discourse_api/api/sso_spec.rb [hunk 9]
```diff
@@ -4,9 +4,51 @@
...
+      'custom.field_1' => 'tomato',
+      # new format
+      custom_fields: {
+        field_2: 'potato'
+      }
+    }
+  end
+  let(:expected_unsigned_payload) do
+    'name=Some+User&username=some_user&email=some%40email.com&'\
+    'avatar_url=https%3A%2F%2Fwww.website.com&external_id=abc&title=ruby'\
```
Comment (vkozyrev) [inline spec/discourse_api/api/sso_spec.rb:31]: note we maintain the custom.custom bug for existing users, new implementations should include `custom_fields` as a hash

**Comment by vkozyrev** [comment]
I think this is ready for you to look at @oblakeerickson 

* I've added `attr_accessor :custom_fields`. This helps with assigning params + testing.
* Added another initialization method that builds `sso` from a Hash
  * This allows sync_sso to automatically work with new attributes, instead of adding them to both the class AND sync_sso method.
* ```
# Custom fields input
{
      ...
      remove_groups: ['c', 'd'],
      ...
      # old format (which results in custom.custom.field_1 in unsigned_payload)
      'custom.field_1' => 'tomato',
      # new format
      custom_fields: {
        field_2: 'potato'
      }
    }
```

File: lib/discourse_api/single_sign_on.rb [hunk 10]
```diff
@@ -15,6 +15,7 @@ class SingleSignOn
     #NONCE_EXPIRY_TIME = 10.minutes # minutes is a rails method and is causing an error. Is this needed in the api?
 
     attr_accessor(*ACCESSORS)
+    attr_accessor :custom_fields
```
Comment (vkozyrev) [inline lib/discourse_api/single_sign_on.rb:18]: Separate to not break `ACCESSORS.each` loops

**Comment by oblakeerickson** [comment]
Great thanks! I’ll review it when I get a chance.

**Pull request merged**